pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -39,284 +108,413 @@ func Add(mgr manager.Manager) error {
+		}
+		r.reqLogger.Errorf("node not found. name %s error %v", request.Name, err)
+		return reconcile.Result{}, err
 	}
 
-	// Create a source for watching noe maintenance events.
-	src := &source.Kind{Type: &kubevirtv1alpha1.NodeMaintenance{}}
+	if node.Annotations == nil {
+		// no annotations on node, nothing to do here.
+		return reconcile.Result{}, nil
+	}
 
-	// Watch for changes to primary resource NodeMaintenance
-	err = c.Watch(src, &handler.EnqueueRequestForObject{}, pred)
+	nodeName := node.ObjectMeta.Name
+
+	// parse status & spec, create initial status object on first call
+	specAnnoData, statusAnnoData, err := r.parseAnnotations(node)
 	if err != nil {
-		return err
+		r.reqLogger.Errorf("invalid value of annotation. name %s error %v", request.Name, err)
+		return reconcile.Result{}, err
 	}
-	return nil
-}
 
-func initDrainer(r *ReconcileNodeMaintenance, config *rest.Config) error {
+	if specAnnoData != nil { // maint. annotation exists. user requests to take the node offline.
 
-	r.drainer = &drain.Helper{}
+		leaseStatus, err, lease := r.doesLeaseExist(r.client, nodeName)
+		switch leaseStatus {
+		case LeaseStatusNotFound:
+			{
+				if err := r.createLease(nodeName, specAnnoData.requestedMaintenenceWindowSeconds); err != nil {
+					statusAnnoData.status = NodeStateFailedToCreateLease
+					_, err = r.setStatusAnnotation(node, statusAnnoData)
+					return reconcile.Result{}, err
+				}
 
-	//Continue even if there are pods not managed by a ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet.
-	//This is required because VirtualMachineInstance pods are not owned by a ReplicaSet or DaemonSet controller.
-	//This means that the drain operation can’t guarantee that the pods being terminated on the target node will get
-	//re-scheduled replacements placed else where in the cluster after the pods are evicted.
-	//KubeVirt has its own controllers which manage the underlying VirtualMachineInstance pods.
-	//Each controller behaves differently to a VirtualMachineInstance being evicted.
-	r.drainer.Force = true
+				statusAnnoData.status = NodeStateCordon
+				if node, err = r.setStatusAnnotation(node, statusAnnoData); err != nil {
+					return reconcile.Result{}, err
+				}
 
-	//Continue even if there are pods using emptyDir (local data that will be deleted when the node is drained).
-	//This is necessary for removing any pod that utilizes an emptyDir volume.
-	//The VirtualMachineInstance Pod does use emptryDir volumes,
-	//however the data in those volumes are ephemeral which means it is safe to delete after termination.
-	r.drainer.DeleteLocalData = true
+			}
 
-	//Ignore DaemonSet-managed pods.
-	//This is required because every node running a VirtualMachineInstance will also be running our helper DaemonSet called virt-handler.
-	//This flag indicates that it is safe to proceed with the eviction and to just ignore DaemonSets.
-	r.drainer.IgnoreAllDaemonSets = true
+		case LeaseStatusActiveOwnedByMeAboutToExpire:
```

Comment:
This only makes sense if we initially set the lease time to a value less than the requested value (but greater than the loop interval and worst case execution time).  If we cannot think of a reason to do that, we should drop this code.

Reply:
deleted it.